### PR TITLE
[5.6] Allow to fetch specific key when using json helper

### PR DESCRIPTION
**Pull request by jerguslejko**
This PR allows to fetch a specific key off of the `TestResponse` class.

Example:

```php
$title = $this->post('/posts', $data)
    ->json('data.title');
```

vs 

```php
$title = $this->post('/posts', $data)
    ->json()['data']['title'];
```

🌵

### CODE REVIEW

File: src/Illuminate/Foundation/Testing/TestResponse.php [hunk 1]
```diff
@@ -500,9 +500,10 @@ public function assertJsonValidationErrors($keys)
     /**
      * Validate and return the decoded response JSON.
      *
+     * @param  string $key
```
Comment (jkniest) [inline src/Illuminate/Foundation/Testing/TestResponse.php]: This should be `@param string|null $key` if I'm not wrong

File: src/Illuminate/Foundation/Testing/TestResponse.php [hunk 2]
```diff
@@ -514,17 +515,18 @@ public function decodeResponseJson()
...
         }
 
-        return $decodedResponse;
+        return data_get($decodedResponse, $key);
     }
 
     /**
      * Validate and return the decoded response JSON.
      *
+     * @param  string $key
```
Comment (jkniest) [inline src/Illuminate/Foundation/Testing/TestResponse.php]: Same as above: This should be @param string|null $key if I'm not wrong

File: src/Illuminate/Foundation/Testing/TestResponse.php (hunk 1 above)
Comment (jerguslejko) [reply src/Illuminate/Foundation/Testing/TestResponse.php]: Oh thanks!

**Comment by mfn** [comment]
If in the old version, `data` or `title` was missing or title, you would get an immediate error (which I consider good).

OTOH with `data_get` you will always receive a value (default value) back and your test breaks at a later point.

Not sure I would consider that an improvement?

Iff it didn't matter for I test I would probably writ `->json()['data']['title'] ?? ''`.

**Comment by sisve** [comment]
A quick reminder that even the TestResponse class is a documented public api. Changing a method signature in a public api is a breaking change. Have you considered targeting the 5.6 release?

**Comment by jkniest** [comment]
But this change, as I see it, is completley optional. You don't need to pass a key in it.

**Comment by jkniest** [comment]
If the $key is null, the normal object will be returned like before.

https://github.com/illuminate/support/blob/master/helpers.php#L454

**Comment by sisve** [comment]
The breaking isn't for those that call the method, but for those that have overridden it. You will get php warnings stating that your model signature does not match the signature that you are overriding.

File: src/Illuminate/Foundation/Testing/TestResponse.php [hunk 3]
```diff
@@ -514,17 +515,18 @@ public function decodeResponseJson()
...
 
-        return $decodedResponse;
+        return data_get($decodedResponse, $key);
     }
 
     /**
      * Validate and return the decoded response JSON.
      *
+     * @param  string|null  $key
      * @return array
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/TestResponse.php]: return type is now wrong

File: src/Illuminate/Foundation/Testing/TestResponse.php [hunk 4]
```diff
@@ -500,9 +500,10 @@ public function assertJsonValidationErrors($keys)
...
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/TestResponse.php]: return type is now wrong

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by sisve** [comment]
Could you change the targeted branch to master to target the 5.6 release since this has breaking changes (changes to method signatures)?

**Comment by jerguslejko** [comment]
Honestly, I do not consider this PR to be a breaking change. I'd like it to be merged into 5.5 so it becomes available right away.

**Comment by devcircus** [comment]
I hate these back-and-forths on breaks, however, whether something is a BC break does not change if you don't agree. It either is or isn't. Technically, if the class isn't declared final, adding a parameter ( even with a default value) is a BC break. If the method is being overwritten (likely), then that app will break. Period.

Should target master, as we shouldn't intentionally break people's apps unless it is an important bug fix(this is not), but that's definitely up to Taylor.

**Pull request merged**